Multi-role and multi-use provisioning tokens

kontsevoy opened the pull request:
## Problem

Teleport was designed to be overly strict when it comes to adding nodes to the cluster. A user would have to generate a token first, then specify this token when launching `teleport` daemon with `node` role. This made it difficult to stand up multiple servers from a single Ansible script, use Vagrant, use AWS AMIs, etc. A less paranoid, static configuration was needed.

Another problem was that Teleport tokens only allowed new nodes to be added to the cluster. There was no way to create a token which would allow a new server to become assume role "node" and a "proxy" role at the same time.

The changes below address both of these issues.
## Configuration changes

`teleport.yaml` now has only one place to specify `auth_token` and it's in the global "teleport" section as opposed to individual "proxy/node" sections. This is possible because the same token can now be used for multiple roles.

Also, the `auth_service` section there is a new setting: `cluster_name`. It will be used in future versions of Teleport to connect trusting clusters together.
## Multi-use Tokens
- Generated tokens can be used multiple times up to TTL.
- By default all tokens have a TTL of 30 minutes, but a new `--ttl` flag for `tctl nodes add` can override that.
- `tctl nodes add` gets a new roles flag, example: `--roles=node,proxy` this allows new nodes to assume both "node" and "proxy" roles.
## Static Tokens

For even less paranoid, Teleport auth now allows to specify a list of static tokens which will always be used to provision new nodes and they never expire.

Consider this teleport `auth` configuration (teleport.yaml):

``` yaml
teleport:
  nodename: auth-a
  log:
    output: stderr
    severity: INFO

auth_service:
  enabled: yes
  cluster_name: cluster-a.local.com
  listen_addr: 0.0.0.0:3025
  # Notice how you can pre-define tokens for adding servers with different roles
  tokens: 
  - "proxy,role:cai8ohV7eem2OhcathuTh4oo"
  - "auth:Hoqu8tai7ailo2PiKoree9no"

ssh_service:
  enabled: no

proxy_service:
  enabled: no
```

The config file for the node (which also runs proxy) which wants to join to this cluster:

``` yaml
teleport:
  nodename: proxy-node
  auth_servers: ["10.0.10.10:3025"]
  auth_token: "cai8ohV7eem2OhcathuTh4oo"

auth_service:
  enabled: no
  listen_addr: 0.0.0.0:3025

ssh_service:
  enabled: yes

proxy_service:
  enabled: yes
```
## Token Management

Because this PR adds multi-use tokens, it also adds the ability to see current/outstanding tokens and allows Teleport admins to delete (revoke) them.

There are two new commands:

```
tctl tokens ls
tctl tokens del
```

These commands work equally well for provisioning (node) tokens as well as signup (user) tokens. They also list "static" tokens. Here's an example of `tctl tokens ls` output which shows one static token (with "never" expiry date) from the yaml file, one user token and one machine token:

```
Token                                Role       Expiry Time (UTC)
-----                                ----       -----------------
696c0471453e75882ff70a761c1a8bfa     Node       17 May 16 03:51 UTC
eoKoh0caiw6weoGupahgh6Wuo7jaTee2     Proxy      never
6fc5545ab78c2ea978caabef9dbd08a5     Signup     17 May 16 04:24 UTC
e
```
## Other Changes

`auth` server is now manually handling TTL, this results in nice and clear error messages. So instead of "invalid parameter" when adding nodes with tokens, users will see:
- XXX cannot join the cluster as 'proxy': token is invalid
- XXX cannot join the cluster as 'proxy': token has expired
- XXX cannot join the cluster as 'proxy': the token does not allow 'proxy' role

Also cleaned up `tctl` --help pages and added a nicer 6-node Vagrant recipe to easier test various deployment scenarios.

Closes #406
Closes #379
Closes #307

kontsevoy committed:
Minor changes to Vagrantfile

kontsevoy committed:
Merge branch 'master' into ev/tmp

kontsevoy committed:
Merge remote-tracking branch 'origin/master' into ev/tmp

kontsevoy committed:
Teleport Host Certificates support multiple roles now

Teleport CA-signed host certificates used to support only one
server role per cert.

This commit adds the ability to store multiple roles in a
certificate, paving the road for multi-role node support in
a near future.

kontsevoy committed:
Spelling in comments

kontsevoy committed:
Added flags to `tctl nodes add`

`--ttl`  : allows to specify TTL for each token
`--roles`: allows to specify which roles the new node will be able to
           assume

kontsevoy committed:
Minor improvement to tunclient

kontsevoy committed:
Vagrant-based 3-node cluster

Each node has only one role.

kontsevoy committed:
Better handling of edge cases in tun client

When a tunnel client (SSH client) finds itself in a situation when all
auth servers are offline, this edge case is handled better.

kontsevoy committed:
Added unit tests to cover role set operations

kontsevoy committed:
Proper handling of foles stored in returned certificates

kontsevoy committed:
Flexible tokens:

- Static never-expiring tokens
- TTL tokens
- Single-use tokens with TTL of 30 mins

kontsevoy committed:
Hooked up static token to node registration

kontsevoy committed:
auth.ValidateToken now understands static tokens

kontsevoy committed:
Minor refactorings and configs for the vagrant environment

kontsevoy commented:
retest this please

klizhentas commented on lib/auth/auth.go:
```diff
@@ -170,19 +168,19 @@ func (a *AuthServer) GetLocalDomain() (string, error) {
 	}, true)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, trace.Errorf("failed to load host CA for '%s': %v", s.DomainName, err)
```
seems like a downgrade since we are loosing error type here. Please return appropriate error here

kontsevoy replied:
The correct way to handle errors is:

- Log the log-level error in-place.
- Form a higher level error and pass it to the caller.

Maybe we should update `trace` package to support this behavior?

In this case I made the change because the function _was_ failing in my real life testing and the low-level message was confusing to a user. The current code delivers superior user experience.
Bubbling up low-level errors up the stack is not helpful to:
a) debug (if you are a developer)
b) change configuration (if you are a user)
Messages like "closed socket" or "bad argument" are only helpful within their context, not outside of it.

klizhentas commented on lib/auth/auth.go:
```diff
@@ -313,27 +303,78 @@ func (s *AuthServer) GenerateServerKeys(hostID string, role teleport.Role) (*Pac
+// a list of roles this token allows its owner to assume, or an error if the token
+// cannot be found
+func (s *AuthServer) ValidateToken(token string) (roles teleport.Roles, e error) {
+	// look at static tokesn first:
```
`tokesn` -> `tokens`

klizhentas commented on lib/auth/auth.go:
```diff
@@ -313,27 +303,78 @@ func (s *AuthServer) GenerateServerKeys(hostID string, role teleport.Role) (*Pac
+	tok, err := s.Provisioner.GetToken(token)
+	if err != nil {
+		log.Warn(err)
+		return nil, trace.Errorf("token not recognized")
```
don't use `trace.Errorf` here. You can make the error originated downards more descriptive, also logging here is unnecessary.

kontsevoy replied:
I have been using teleport in real life and this behavior is user-facing and required. All my changes to errors and logging were made to improve Teleport Administrator experience, namely:

- CLI error messages
- Log output

In both cases I was trying to make it obvious to a user to diagnose and fix configuration errors.

klizhentas commented on lib/auth/auth.go:
```diff
@@ -313,27 +303,78 @@ func (s *AuthServer) GenerateServerKeys(hostID string, role teleport.Role) (*Pac
+	// look at the tokens in the token storage
+	tok, err := s.Provisioner.GetToken(token)
+	if err != nil {
+		log.Warn(err)
```
logging here is unnecessary

kontsevoy replied:
removed

klizhentas commented on lib/auth/auth.go:
```diff
@@ -313,27 +303,78 @@ func (s *AuthServer) GenerateServerKeys(hostID string, role teleport.Role) (*Pac
+	now := s.clock.Now().UTC()
+	if tok.Expires.Before(now) {
+		if err = s.DeleteToken(token); err != nil {
+			log.Error(err)
```
this can be `NotFoundError` if token has expired, this log line is unnecessary

klizhentas commented on lib/auth/auth.go:
```diff
@@ -313,27 +303,78 @@ func (s *AuthServer) GenerateServerKeys(hostID string, role teleport.Role) (*Pac
+
+// enforceTokenTTL deletes the given token if it's TTL is over. Returns 'false'
+// if this token cannot be used
+func (s *AuthServer) checkTokenTTL(token string) bool {
```
in general checkers in teleport return `error` if something goes wrong or `nil` if check has succeeded. This will spare you from logging errors everywhere, since you simply return them.

klizhentas commented on lib/auth/init.go:
```diff
@@ -71,14 +71,30 @@ type InitConfig struct {
+	StaticTokens []StaticToken
+}
+
+// StaticToken describes a pre-defined always-live token, usually stored in a config file.
```
I think `StaticToken` should be named `ProvisioningToken` as it's not clear what `Static` means however this token is used for `Provisioning` a new server

klizhentas replied:
see even in your comment above you refer to them as `provisioning token`

klizhentas commented on lib/auth/tun.go:
```diff
@@ -721,11 +724,15 @@ func (c *TunClient) setAuthServers(servers []utils.NetAddr) {
+	// see if we have any auth servers online:
+	authServers := c.getAuthServers()
+	if len(authServers) == 0 {
+		return nil, trace.Errorf("all auth servers are offline")
```
don't use `trace.Errorf`

klizhentas commented on lib/auth/tun.go:
```diff
@@ -734,22 +741,25 @@ func (c *TunClient) getClient() (*ssh.Client, error) {
-		log.Infof("TunDialer: ssh.Dial: %v", err)
-
+	const dialRetryInterval = time.Duration(time.Millisecond * 50)
+	for attempt := 0; attempt < 5; attempt++ {
```
magic constant here. Let's move this to `defaults`

klizhentas commented on lib/auth/tun.go:
```diff
@@ -734,22 +741,25 @@ func (c *TunClient) getClient() (*ssh.Client, error) {
-	if err != nil {
-		log.Infof("TunDialer: ssh.Dial: %v", err)
-
+	const dialRetryInterval = time.Duration(time.Millisecond * 50)
```
magic constant here, let's move it to `defaults`

klizhentas replied:
also, I think that `50` ms is too aggressive. Over 5 attempts is't just quarter of a second worth of retries, usually network issues don't get resolved in such a short period

klizhentas commented:
This PR misses documentation, please update docs as well.

klizhentas commented:
I also think that we need to provide a clear workflow on how to:

* list existing tokens
* rotate tokens periodically
* revoke tokens in case if something goes wrong

I think this all deserves 

```
tctl tokens ls
                 rm
                 add
```

and we should document this as well.

klizhentas commented:
I also liked your vagrant file sample, let's include it int he docs alongside with ansible example.

kontsevoy committed:
Documentation edits

- Updated docs for "advertise_ip"
- Updated docs for "tokens" in teleport.yaml
- Updated "adding nodes" section in the admin guide and in the
  quickstart guide
- Crated `run-docs` make target for convenient live documentation
  editing.

kontsevoy commented:
Updated the documentation, see #643cd5c
Added some missing tests to cover configuration.

Token invalidation will come later. It's a P2 enhancement.

kontsevoy commented:
I do not think your suggestions for handling errors are helpful. The reason these changes were made was to improve end-user experience. Teleport in general has a tendency to ignore errors: it bubbles up the low level errors (like "file does not exist" or "closed socket") all the way to the UI.

A user is then left without any helpful information whatsoever. Trying to add a node? "Closed socket!" Trying to list servers? "too many open files", etc. From a end-user perspective this is about as valuable as not handling errors at all.

These changes aim to:
- Log a _helpful_ message to Teleport log for an admin.
- Send a human-readable error via the wire to the calling client.

All of these changes were forced by real-world usability testing. 

I am not saying Teleport has a faulty error handling. I am saying that Teleport error handling is optimized for SaaS-like environment when errors are shipped into an external structured log and analyzed later by developers. 

This approach does not provide an acceptable user experience for installable  software like Teleport which "speaks" via CLI errors and user-readable log messages (and developers aren't around)

kontsevoy committed:
Added tctl CLI commands for listing and deleting tokens

TODO: hook them up to server-side implementation

kontsevoy committed:
Back-end implementation of token management

- User tokens (signup tokens) and node nodes (provisioning tokens) are
  managed via the same API calls.

- User tokens are converted to machine tokens (with Signup role)

- Static node tokens have "Expiry" date of Unix(0) i.e. Jan 1, 1970

kontsevoy commented:
Added "Token Management":

## Token Management

Because this PR adds multi-use tokens, it also adds the ability to see current/outstanding tokens and allows Teleport admins to delete (revoke) them.

There are two new commands:
```
tctl tokens ls
tctl tokens del
```
These commands work equally well for provisioning (node) tokens as well as signup (user) tokens. They also list "static" tokens. Here's an example of `tctl tokens ls` output which shows one static token (with "never" expiry date) from the yaml file, one user token and one machine token:

```
Token                                Role       Expiry Time (UTC)
-----                                ----       -----------------
696c0471453e75882ff70a761c1a8bfa     Node       17 May 16 03:51 UTC
eoKoh0caiw6weoGupahgh6Wuo7jaTee2     Proxy      never
6fc5545ab78c2ea978caabef9dbd08a5     Signup     17 May 16 04:24 UTC
e
```

kontsevoy committed:
Added a nicer error message ...

for when a user tries to delete a statically configured token

kontsevoy committed:
Fixed tests

kontsevoy committed:
Updated documentation

Added "Revoking Invitations" to the admin guide.

klizhentas commented on docs/admin-guide.md:
````diff
@@ -329,6 +376,37 @@ turing        d52527f9-b260-41d0-bb5a-e23b0cfe0f8f     10.1.0.5:3022      distro
+6fc5545ab78c2ea978caabef9dbd08a5     Signup     17 May 16 04:24 UTC
+```
+
+In this exampe, wthe first token with "never" expiry date is a static token configured via
````
example

klizhentas commented:
k, you are good to merge

kontsevoy committed:
Merge remote-tracking branch 'origin/master' into ev/multi-role

Conflicts:
	lib/auth/tun.go
